Support Hashes convertible to JSON on post, put and patch requests

quimrstorres opened the pull request:
The initial idea was to simply fallback to an object that responds to
`#to_json`, but the `#to_json` extensions implemented in Ruby's JSON
module do not always produce the intended JSON. Examples:

```
"{\"foo\": \"bar\"}".to_json
=> "\"{\\\"foo\\\": \\\"bar\\\"}\""

nil.to_json
=> "null"
```

For this reason, I've decided to only call the method on `Hash objects`.

pezra commented on lib/hal_client.rb:
```diff
@@ -86,6 +86,8 @@ def def_unsafe_request(method)
 
         req_body = if data.respond_to? :to_hal
                      data.to_hal
+                   elsif data.is_a?(Hash) && data.respond_to?(:to_json)
```
we we need the `respond_to?` check?

quimrstorres replied:
No, it is not needed. :+1: Will fix!

pezra commented:
Mind adding a spec for this behavior. Other than that it looks good.

quimrstorres commented:
@pezra I've fixed things based on your comments and added the missing spec.

I looked into the problem with the build and it seems to be a known issue with RVM (rvm/rvm#3515) on Ubuntu.

pezra commented:
I bumped the version of rubinius on the master branch. Mind rebasing to
make sure this branch still passes? Thanks.

On Mon, Dec 21, 2015 at 9:58 AM, quimrstorres <notifications@github.com>
wrote:

> @pezra <https://github.com/pezra> I've fixed things based on your
> comments and added the missing spec.
>
> I looked into the problem with the build and it seems to be a known issue
> with RVM (rvm/rvm#3515 <https://github.com/rvm/rvm/issues/3515>) on
> Ubuntu.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pezra/hal-client/pull/43#issuecomment-166359198>.
>

quimrstorres commented:
Done! Let me know if there's something else you think should be done differently.

pezra commented:
Thanks for the improvement, @quimrstorres!

pezra commented:
FYI, this is release (in the 3.12.0 version of the gem).

quimrstorres commented:
Awesome! Thanks @pezra
